### DISCUSSION TITLE
Started adding demos

### DISCUSSION THREAD

**Discussion by martinvonz** [jj-vcs/jj#37] [Show and tell]
I've started adding demos to the README. I copied the automation from https://github.com/arxanas/git-branchless/tree/master/demos so thank you so much for writing that code, @arxanas! I had some trouble getting it to not use the `$PS1` from my Bash configs. I fixed that by doing [this](https://github.com/martinvonz/jj/blob/234cb15ff897b055a31aea6f6dad8de38fb16b4b/demos/demo_helpers.sh#L48-L53). Perhaps you'll want to do something similar.

**Comment by martinvonz** [comment]
I've added several demos to the README now. I'd appreciate feedback on all aspects: content, speed, workflow, etc.

I'll add more demos of conflict handling and automatic rebase later. Maybe demos of revsets too.

**Reply by arxanas** [reply]
* Try to keep your explanatory messages a lot shorter per line, so that they can be viewed on mobile.
* I didn't realize for a couple of minutes that the current entry in the commit graph is indicated by green text. The `@` is somewhat subtle, and it's how I figured it out. Can you make this more obvious?
* There's a typo "id changed" somewhere.
* I missed the demos when skimming through the README because they're minimized. I think peoples' attention is drawn to pictures, and they're still easy to skip if one decides it's not relevant.
* Lately I've been of the opinion that if the demo is more than a few seconds, maybe it's not worth doing. I want to adjust my `git undo` demo to be like five seconds long, just with a bad merge/rebase and then running a non-interactive `git undo`. Then I can also embed it directly in the README as an autoplaying SVG. (I tried that once, but it was like 5 MB at the time due to the demo length, so it took forever to process.)
* I generally liked the pacing in the demos. The somewhat fast explanatory text accompanied by a longer ending pause after the command seems like a better setup than what I used. I read very quickly, so make sure to run your demos by a slow reader to ensure they're still accessible.

**Reply by arxanas** [reply]
This was the mobile view I had on Asciinema

![Screenshot_20211027-125050_Firefox.jpg](https://user-images.githubusercontent.com/454057/139137959-0c87388d-a620-4c27-bd1f-08e3c1f6008f.jpg)

**Reply by arxanas** [reply]
Could you add color to the `jj status` output? My eye was not drawn to the output like I think it should have been. I think the trick is to color the important variable material and keep any explanatory text in the default color.

**Reply by arxanas** [reply]
There's a lot of sequential text. It might be a good idea to add a newline in between different sections, just to make it easier to pick out the boundaries of the logical operations. Like writing files and checking status and committing(/not explicitly committing) is all one operation.

**Reply by martinvonz** [reply]
> Try to keep your explanatory messages a lot shorter per line, so that they can be viewed on mobile.

Good point, I'll go through and rewrap (and maybe shorten if I can).

> the current entry in the commit graph is indicated by green text

That's actually not what the green text indicates :) The green text indicates that changes you make in the working copy will update the commit (if you have it checked out).

> The @ is somewhat subtle [...]. Can you make this more obvious?

In my hg setup, I actually use orange background color for the description to highlight it. That won't work well with jj, since the description is often empty. I'll send patches for making the text bright instead. Thanks for the suggestion.

> Could you add color to the jj status output?

Ah, I've meant to do that for a while. Done (in https://github.com/martinvonz/jj/commit/452fecb7c4f60b4522c3322bf07969d9536fb9fb) :)

I'll take a look at the rest of your comments soon. Thanks!

**Reply by martinvonz** [reply]
> There's a typo "id changed" somewhere.

I think that was intentional ("commit id changed") but I've replaced "id" by the more correct "ID" throughout now :)

> I missed the demos when skimming through the README because they're minimized.

Fixed.

> There's a lot of sequential text. It might be a good idea to add a newline in between different sections

Done. Hopefully it looks better now.

> Lately I've been of the opinion that if the demo is more than a few seconds, maybe it's not worth doing.

Interesting. I guess the idea is to have the demos just spark interest, not to explain how things work. That's a larger change. I'll think more about it and will probably update the demos later. I was planning on adding more demos anyway (e.g. for juggling conflicts, revsets, automatic rebase).

**Reply by martinvonz** [reply]
Unfortunately the asciinema web player seems to use bold instead of bright colors, which makes the highlighting of the checked-out commit much less clear in the demos :(

**Reply by arxanas** [reply]
Nice, they're looking a lot better.

> I think that was intentional ("commit id changed") but I've replaced "id" by the more correct "ID" throughout now :)

Heh, I thought that was supposed to be "commit is changed".

My personal taste would be to liberally use `jj log` after every operation. In https://asciinema.org/a/uu2yF0FzPiRigbYQYok2PHmX5, surely it would be good to show how the working copy is committed and somewhere else in the graph as soon as you check out away from it? Then again, that might just be because I'm now heavily dependent on viewing the commit graph to do my work...

I see now that the bright rendering isn't so great in Asciinema. It looks like there's a selection of five themes if you go to Edit » Settings, so maybe one of them renders better? I think the only advantage `git-branchless` has in its graph output is that the current checkout is a filled in glyph, and the rest are hollow:

```
$ git sl
⋮
◇ 2bbb2f12 17d Create foo
┣━┓
┃ ◯ 0128469f 4m (work) Update foo locally
┃
◆ 6effb47f 2d (master) Update foo remotely  
```

**Reply by martinvonz** [reply]
> My personal taste would be to liberally use jj log after every operation. In https://asciinema.org/a/uu2yF0FzPiRigbYQYok2PHmX5, surely it would be good to show how the working copy is committed and somewhere else in the graph as soon as you check out away from it?

Oh, I definitely agree! I'll fix that. When demoing this manually, I like to have a split tmux window with `watch -n 0.1 jj log` running in one of the panes. That makes it quite clear what the effects of any changes in the working copy and any commands are.

> t looks like there's a selection of five themes if you go to Edit » Settings, so maybe one of them renders better?

Nice! I changed my default to the Tango theme. That seems to have a much clearer difference between bright and regular colors.

**Reply by arxanas** [reply]
I was able to follow the demo pretty clearly now that you added the `log` more often and switched the color scheme. I think I have to go switch the color scheme for my demos now too...
